src/k2/tso/service/Clock.h

```diff
@@ -0,0 +1,152 @@
+    bool operator==(const GPSTimePoint& o) const;
+};
+
+// The GPS clock can be used to obtain timestamps from a GPS clock
+// Note that this class internally operates in terms of nanoseconds-since-epoch instead
+// of using std::chrono datatypes since the stl datatypes are not volatile-friendly
+class GPSClock {
+   public:
+    // Generates GPS timepoints. These are guaranteed to be monotonically increasing.
+    static GPSTimePoint now();
+
+    // Should be called often to poll for new GPS clock values.
+    // The assumption here is that the GPS hardware will be able to DMA a value in some memory
+    // location, which we can poll here and associate with a steady clock value ASAP.
+    static void poll();
+
+    // busy-loop poll to detect at least one update to the gps clock value within a reasonable error
+    // This method should be called before the application proceeds with its startup (i.e. in main())
+    static void initialize();
+
+    // the base error tolerance for the gps device
+    // TODO this should come based on particular manufacturer guarantees and calibrated values
+    static constexpr inline uint32_t BASE_ERROR_NANOS = 1'000;
+    // sentinel value to indicate infinite drift in nanoseconds. If this value is returned, the drift should
+    // be considered infinite
+    static constexpr inline uint32_t INF_DRIFT = std::numeric_limits<uint32_t>::max();
+
+    // TODO, this should be benchmarked and confirmed for the particular motherboard crystal.
+    // Standard drift  is ~30 PPM on desktop hardware (or 30us per second). We could achieve better
+    // with server-grade components with better quality crystals (e.g. TCXO)
+    // Some published numbers
+    //
+    // 2021 IBM power8/9 servers with 3sec/day which is ~35PPM
+    //   https://www.ibm.com/support/pages/time-drift-power8-and-power9-servers
+    //
+    // TCXO-based oscillators can achieve 1us/day (1.2PPM)
+    //  - https://www.gsc-europa.eu/sites/default/files/sites/all/files/Report_on_User_Needs_and_Requirements_Timing_Synchronisation.pdf
+    // - https://www.microsemi.com/product-directory/high-reliability-rugged-oscillators/4848-tcxo
+    static constexpr inline uint32_t DRIFT = 30;  // drift rate as parts-per DRIFT_PERIOD
+    static constexpr inline uint32_t DRIFT_PERIOD = 1'000'000;
+
+    // get the drift over a particular duration
+    // note the possibility for rounding errors here due to integer math -
+    //      Accumulating results for small increments is not advisable
+    static uint32_t getDriftForDuration(uint64_t durationNanos);
+
+private:
+    // obtain gps and steady timestamps
+    static uint64_t _gpsNow();
+    static uint64_t _steadyNow();
+    // Holds the same information as a GPSTimePoint, but in a volatile-friendly data members
+    struct _GPSNanos {
+        _GPSNanos();
+        _GPSNanos(uint64_t steady, uint64_t real, uint32_t error);
+        void operator=(const volatile _GPSNanos& o) volatile;
```

Comment:
Do we really want these methods declared volatile? My understanding is that it will treat all member variables as volatile within the functions. I'm not sure we always want that.